scheduling/dispatcher.rb

```diff
@@ -1,93 +1,228 @@
 # frozen_string_literal: true
 
+# Scheduling::Dispatcher is used for running strands. It finds strands
+# that should be run, and runs them using a thread pool. It keeps a
+# record of which strands are currently being processed, so that it
+# does not retrieve them.
 class Scheduling::Dispatcher
-  attr_reader :notifiers, :shutting_down
+  attr_reader :shutting_down
 
-  def initialize
-    @apoptosis_timeout = Strand::LEASE_EXPIRATION - 29
-    @notifiers = []
+  STRAND_RUNTIME = Strand::LEASE_EXPIRATION / 4
+  APOPTOSIS_MUTEX = Mutex.new
+
+  def initialize(apoptosis_timeout: Strand::LEASE_EXPIRATION - 29)
     @shutting_down = false
+
+    # How long to wait in seconds from the start of strand run
+    # for the strand to finish until killing the process.
+    @apoptosis_timeout = apoptosis_timeout
+
+    # Hash of currently executing strands.  This can be accessed by
+    # multiple threads concurrently, so access is protected via a mutex.
+    @current_strands = {}
+    # Mutex for current strands
+    @mutex = Mutex.new
+
+    # The thread pool size.  Set to one less than the number of database
+    # connections, to ensure there is a connection always available for the
+    # main thread, for the query to find strands to run.
+    pool_size = Config.db_pool - 1
+
+    # The Queue that all threads in the thread pool pull from.  This is a
+    # SizedQueue to allow for backoff in the case that the thread pool cannot
+    # process jobs fast enough. When the queue is full, the main thread to
+    # push strands into the queue blocks until the queue is no longer full.
+    # The queue size is 4 times the size of the thread pool, as that should
+    # ensure that for a busy thread pool, there are always strands to run.
+    # This should only cause issues if the thread pool can process more than
+    # 4 times its size in the time it takes the main thread to refill the queue.
+    @strand_queue = SizedQueue.new(pool_size * 4)
+
+    # An array of thread pool data.  This starts pool_size * 2 threads.  Half
+    # of the threads are strand threads, responsible for running the strands.
+    # The other half are apoptosis threads, each responsible for monitoring
+    # the related strand thread.  The strand thread notifies the apoptosis
+    # thread via a queue when it starts, and also when it finishes.  If a
+    # strand thread does not notify the apoptosis thread of it finishing
+    # before the apoptosis timeout, then the process exits.
+    @thread_data = Array.new(pool_size) do
+      start_thread_pair(@strand_queue)
+    end
+
+    # A prepared statement to get the strands to run.  A prepared statement
+    # is used to reduce parsing/planning work in the database.
+    @strand_ps = Strand
+      .where(
+        Sequel.|({lease: nil}, Sequel[:lease] < Sequel::CURRENT_TIMESTAMP) &
+        (Sequel[:schedule] < Sequel::CURRENT_TIMESTAMP) &
```

Comment:
A simple approach to deal with the 1 sec sleep delays could be to slightly relax this condition and include strands whose schedule will be due soon (in 1 sec or less). My understanding is that our prog should be ok with being executed slightly earlier than requested. The `order_by` makes sure that strands that are already due have higher priority. But I see that this has relatively far reaching semantic consequences and could introduce unexpected behavior.